From 731e0296d566c0fb8af0c965586d8b90dcdcc1f2 Mon Sep 17 00:00:00 2001
From: Dave Airlie <airlied@redhat.com>
Date: Fri, 12 Sep 2014 01:56:51 -0400
Subject: [drm] i915: lock around link status and link training

Message-id: <1410487019-24661-14-git-send-email-airlied@redhat.com>
Patchwork-id: 93922
O-Subject: [RHEL7.1 BZ1140440 PATCH 13/21] drm/i915: lock around link status and link training.
Bugzilla: 1140440
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Rob Clark <rclark@redhat.com>

bugzilla: 1140440
upstream: 5b215bcff50d549d73e43c09bcccf8eebcc95bac

We need to take the connection mutex around the link status
check for non-MST case, but also around the MST link training
on short HPDs.

I suspect we actually should have a dpcd lock in the future as
well, that just lock the local copies of dpcd and flags stored
from that.

Reviewed-by: Daniel Vetter <daniel.vetter@ffwll.ch>
Signed-off-by: Dave Airlie <airlied@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/i915/intel_dp.c b/drivers/gpu/drm/i915/intel_dp.c
index c0db398..25ed7b0 100644
--- a/drivers/gpu/drm/i915/intel_dp.c
+++ b/drivers/gpu/drm/i915/intel_dp.c
@@ -3400,11 +3400,13 @@ intel_dp_check_mst_status(struct intel_dp *intel_dp)
 void
 intel_dp_check_link_status(struct intel_dp *intel_dp)
 {
+ struct drm_device *dev = intel_dp_to_dev(intel_dp);
  struct intel_encoder *intel_encoder = &dp_to_dig_port(intel_dp)->base;
  u8 sink_irq_vector;
  u8 link_status[DP_LINK_STATUS_SIZE];
 
- /* FIXME: This access isn't protected by any locks. */
+ WARN_ON(!drm_modeset_is_locked(&dev->mode_config.connection_mutex));
+
  if (!intel_encoder->connectors_active)
   return;
 
@@ -3928,7 +3930,9 @@ intel_dp_hpd_pulse(struct intel_digital_port *intel_dig_port, bool long_hpd)
     * we'll check the link status via the normal hot plug path later -
     * but for short hpds we should check it now
     */
+   drm_modeset_lock(&dev->mode_config.connection_mutex, NULL);
    intel_dp_check_link_status(intel_dp);
+   drm_modeset_unlock(&dev->mode_config.connection_mutex);
   }
  }
  return false;
-- 
1.7.1